File: palsquare.py
"""
ID: totheso1
LANG: PYTHON3
TASK: palsquare
"""
import operator
import itertools

fin = open("palsquare.in", "r")
fout = open("palsquare.out", "w")


def fprint(*args, **kwargs):
    print(*args, file=fout, **kwargs)


# Code start
digits = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"


def base_repr(number: int, base: int = 2, padding: int = 0) -> str:
    # if base > len(digits):
    #     raise ValueError("Bases greater than 36 not handled in base_repr.")
    # elif base < 2:
    #     raise ValueError("Bases less than 2 not handled in base_repr.")

    num = abs(number)
    res = []
    while num:
        res.append(digits[num % base])
        num //= base
    if padding:
        res.append("0" * padding)
    if number < 0:
        res.append("-")
    return "".join(reversed(res or "0"))


BASE = int(fin.read())
for num in range(1, 301):
    square = base_repr(num * num, BASE)
    if all(map(operator.eq, itertools.islice(square, len(square) >> 1), reversed(square))):
        fprint(base_repr(num, BASE), square)
# Code end

File: palsquare.cpp
/*
ID: totheso1
LANG: C++14
TASK: palsquare
*/

#include <algorithm>
#include <cstdlib>
#include <fstream>
#include <iterator>
#include <string>

using namespace std;

static ifstream fin("palsquare.in");
static ofstream fout("palsquare.out");

template <typename T>
T fin_get() {
  return *istream_iterator<T>(fin);
}

template <typename C>
C fin_get_collection(int size) {
  C ret;
  copy_n(istream_iterator<typename C::value_type>(fin), size,
         back_inserter(ret));
  return ret;
}

static string base_repr(int number, const int base = 2) {
  static const string digits = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";

  string res;
  while (number) {
    res.push_back(digits[number % base]);
    number /= base;
  }
  reverse(begin(res), end(res));
  return res;
}

static bool is_palindrome(const string &s) {
  auto rb = crbegin(s);
  for (auto b = cbegin(s), e = b + (s.size() >> 1); b != e; ++b, ++rb) {
    if (*b != *rb) {
      return false;
    }
  }
  return true;
}

int main() {
  const int BASE = fin_get<int>();
  for (int num = 1; num <= 300; ++num) {
    const string square = base_repr(num * num, BASE);
    if (is_palindrome(square)) {
      fout << base_repr(num, BASE) << ' ' << square << endl;
    }
  }
}
